pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp

```diff
@@ -632,5 +636,128 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
+     * @brief Set the seed of the internal random generator
+     *
+     * @param seed Seed
+     */
+    void setSeed(const size_t seed) { rng.seed(seed); }
+
+    /**
+     * @brief Sample 0 or 1 for given probabilities
+     *
+     * @param probs Probabilities of 0.
+     */
+    auto randomSample(const PrecisionT prob_0) {
+        std::discrete_distribution<int> distrib({prob_0, 1. - prob_0});
+        return distrib(rng);
+    }
+
+    /**
+     * @brief Calculate probabilities of measuring 0 or 1 on a specific wire
+     *
+     * @param wire Wire
+     * @return std::vector<PrecisionT> Both probabilities of getting 0 or 1
+     */
+    auto probs(const std::size_t wire) -> std::vector<PrecisionT> {
+        // TODO: use Measurements.probs() instead (resolve circular dependency)
+        auto *arr = this->getData();
+        const size_t stride = pow(2, this->num_qubits_ - (1 + wire));
+        const size_t vec_size = pow(2, this->num_qubits_);
+        const auto section_size = vec_size / stride;
+        const auto half_section_size = section_size / 2;
+
+        PrecisionT prob_0{0.};
+        // zero half the entries
+        // the "half" entries depend on the stride
+        // *_*_*_*_ for stride 1
+        // **__**__ for stride 2
+        // ****____ for stride 4
+        for (size_t idx = 0; idx < half_section_size; idx++) {
+            const size_t offset = stride * 2 * idx;
+            for (size_t ids = 0; ids < stride; ids++) {
+                prob_0 += std::norm(arr[offset + ids]);
+            }
+        }
+
+        std::vector<PrecisionT> probs{prob_0, PrecisionT(1.) - prob_0};
+        return probs;
+    }
+
+    /**
+     * @brief Collapse the state vector after having measured one of the
+     * qubits
+     *
+     * @param wire Wire to collapse.
+     * @param branch Branch 0 or 1.
+     */
+    void collapse(const std::size_t wire, const bool branch) {
+        auto *arr = this->getData();
+        const size_t stride = pow(2, this->num_qubits_ - (1 + wire));
+        const size_t vec_size = pow(2, this->num_qubits_);
+        const auto section_size = vec_size / stride;
+        const auto half_section_size = section_size / 2;
+
+        // zero half the entries
+        // the "half" entries depend on the stride
+        // *_*_*_*_ for stride 1
+        // **__**__ for stride 2
+        // ****____ for stride 4
+        const size_t k = branch ? 0 : 1;
+        for (size_t idx = 0; idx < half_section_size; idx++) {
+            const size_t offset = stride * (k + 2 * idx);
+            for (size_t ids = 0; ids < stride; ids++) {
+                arr[offset + ids] = {0., 0.};
+            }
+        }
+
+        normalize();
+    }
+
+    /**
+     * @brief Normalize vector (to have norm 1).
+     */
+    void normalize() {
+        auto *arr = this->getData();
+        PrecisionT norm = std::sqrt(squaredNorm(arr, this->getLength()));
+
+        if (norm > std::numeric_limits<PrecisionT>::epsilon() * 1e2) {
```

Comment:
In this `normalize` function, I just want to make sure we don't divide by zero.

- If the norm is zero in our mid-measure scenario, the collapse was done by trying to keep elements of the vector that already had zero amplitude. 
- In the dynamic shots approach, as we sample from the probabilities, this case will not occur.
- In the fixed shots approach (catalyst), this case may occur. But we can avoid this if catalyst checks the probabilities first and stops in case probability of getting `branch` is zero.

Reply:
Please check with the catalyst team the best course of action here, but if this is not needed I'm in favor of removing it.

Reply:
The analogy is usually give is this is that projecting the state on a zero-probability branch is the quantum programming equivalent of a division by zero. It thus requires special handling at either compile time or runtime as in classical programming.

Reply:
It looks like right now if the probabilities are very small, the normalization is just skipped. Is that better than having a statevector with infinities? 🤔

Reply:
This should not happen with the revised workflow, so @tomlqc how about simply throwing an error with `PL_ABORT_IF`?

Reply:
I'm just unsure about `std::numeric_limits<PrecisionT>::epsilon() * 1e2`. The `1e2` is arbitrary (I've searched for similar usage in lightning and found some 1e2 or 1e4 at other places). 
Does everyone agree with `1e2` and `PL_ABORT_IF` ?

Reply:
Thank you for clarifying, we can keep it like that, but possibly throwing an error as suggested by @vincentmr.